src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml

```diff
@@ -0,0 +1,243 @@
+  </struct>
+
+
+
+  <!-- Enum definitions for select locations command response-->
+  <enum name="SelectLocationsStatus" type="enum8">
+    <cluster code="0x0150"/>
+    <item value="0x00" name="Success"/>
+    <item value="0x01" name="UnsupportedLocation"/>
+    <item value="0x02" name="DuplicatedLocations"/>
+    <item value="0x03" name="InvalidInMode"/>
+    <item value="0x04" name="InvalidSet"/>
+  </enum>
+
+  <!-- Enum definitions for skip location command response-->
+  <enum name="SkipCurrentStatus" type="enum8">
+    <cluster code="0x0150"/>
+    <item value="0x00" name="Success"/>
+    <item value="0x01" name="InvalidLocation"/>
+    <item value="0x02" name="InvalidInMode"/>
+  </enum>
+
+
+  <cluster>
+    <domain>General</domain>
+    <name>Service Area</name>
+      <code>0x0150</code>
+    <define>SERVICE_AREA_CLUSTER</define>
+    <description>The Service Area cluster provides an interface for controlling the locations where a device should operate, and for querying the current location.</description>
+
+    <!-- Attributes -->
+    <attribute side="server" code="0x0000" define="SupportedLocations"  type="array"     entryType="LocStruct"  writable="false" isNullable="false" optional="false">SupportedLocations</attribute>
+    <attribute side="server" code="0x0001" define="SupportedMaps"       type="array"     entryType="MapStruct"  writable="false" isNullable="false" optional="false">SupportedMaps</attribute>
+    <attribute side="server" code="0x0002" define="SelectedLocations"   type="array"     entryType="int32u"     writable="false" isNullable="true"  optional="false" scene="true">SelectedLocations</attribute>
+    <attribute side="server" code="0x0003" define="CurrentLocation"     type="int32u"                           writable="false" isNullable="true"  optional="true">CurrentLocation</attribute>
+    <attribute side="server" code="0x0004" define="EstimatedEndTime"    type="epoch_s"                          writable="false" isNullable="true"  optional="true" >EstimatedEndTime</attribute>
+    <attribute side="server" code="0x0005" define="Progress"            type="array"     entryType="ProgStruct" writable="false" isNullable="true"  optional="true">Progress</attribute>
+
+    <!-- Commands -->
+    <command source="client" code="0x00" name="SelectLocations" response="SelectLocationsResponse" optional="false">
+      <description>
+        Command used to select a set of device locations, where the device is to operate
+      </description>
+      <arg name="NewLocations" type="int32u" array="true"/>
+    </command>
+
+    <command source="server" code="0x01" name="SelectLocationsResponse" disableDefaultResponse="true" optional="false">
+      <description>
+        This command is sent by the device on receipt of the SelectLocations command.
+      </description>
+      <arg name="Status"     type="SelectLocationsStatus"               optional="false"/>
```

Comment:
I wonder if the type here has to be `enum8`, as defined in the spec. This depends on whether there is the possibility for us to want to add values form a different enum8, like vendor defined or a derived cluster.

Reply:
The spec says enum8, so unless we change the spec we should use that.

Should we change the spec?  (TL:DR - I don't know.)

The StatusText field does allow for vendor specific error information.

In operational-state-cluster.xml, which has derived clusters, the enum type is used for the type.
<attribute side="server" code="0x0004" define="OPERATIONAL_STATE"       type="OperationalStateEnum" 
However, the base server code deals with the state as a raw unsigned value.
uint8_t mOperationalState    (operational-state-server.h)

In the derived operational-state-rvc-cluster, additional states are defined, and the attribute is defined with the generic.
<attribute side="server" code="0x0004" define="OPERATIONAL_STATE"       type="enum8"

So there is at least one example where the base cluster definition uses the specific  type rather than the generic enum8 for an extensible value.  But it's not clear to me what the advantage is of the specific type in the base xml definition, since it isn't used in the server code. (Generated code for JSON conversion? Logging? - but then this isn't extended to derived classes.)

Reply:
There is also a GenericOperationalState defined in operational-state-cluster-objects.h which provides extensibility of the operational state values by associating a text label with specific enum values. 

These are stored in mOperationalStateList for specific derived implementations to provide definitions of all supported enum values - both standard and derived class specific - as in rvc-operational-state-delegate.h.

If we want to fully support derived classes having additional device specific state, we should probably add something similar.

Reply:
I don't think we need to worry about supporting derived clusters at this stage except to consider if the way we're doing things will inhibit their addition in the future.

I also prefer to use strong types where possible. In this case, if we use the actual type, only that type will be accepted in the base implementation of the cluster. Hence, if this cluster is derived and if the derived version of this cluster would like to have derived cluster specific additions to this enum, they would be of a different type, as they would be in a different namespace, and therefor not accepted by the base cluster. To do this, we would also need to require that certain number spaces of the enum are reserved for derived clusters.

At this stage, the question we want to answer is, if we used the actual type, would we be able to change this to an enum8 in the future? I think that the answer to this question is yes.

If there is such a change, the SDK API of functions would change, requiring an enum8 instead of the actual type. This change would be something that needs to be implemented by the vendors that use this cluster when they choose to use the later version of the SDK. However, since this is not a change to the data on the wire (the data being sent back and fort would still be a uint8), we can argue that this would not be a breaking change as clients and devices using the different versions of the SDK would still be able to communicate.

## TLDR

Given that currently we do not see the future need of clusters derived form this one, we can go ahead and use the actual types as changing this to an enum8 in the future is not a breaking change.